backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java

```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
+      currentOrgGroupMapForUser.values().forEach(g -> g.removeUser(user.getId()));
       throw new IllegalGraphqlArgumentException(
           "Cannot update privileges of Okta user in organization they do not belong to.");
     }
 
-    // Add user to new groups
-    Set<String> groupNamesToAdd = new HashSet<>();
-    groupNamesToAdd.addAll(
-        roles.stream()
-            .filter(r -> r != OrganizationRole.getDefault())
-            .map(r -> generateRoleGroupName(orgId, r))
-            .collect(Collectors.toSet()));
+    Set<String> expectedOrgGroupNamesForUser = new HashSet<>();
+    expectedOrgGroupNamesForUser.add(groupOrgDefaultName);
+    expectedOrgGroupNamesForUser.addAll(
+        roles.stream().map(r -> generateRoleGroupName(orgId, r)).collect(Collectors.toSet()));
     if (!PermissionHolder.grantsAllFacilityAccess(roles)) {
-      groupNamesToAdd.addAll(
+      expectedOrgGroupNamesForUser.addAll(
           facilities.stream()
               .map(f -> generateFacilityGroupName(orgId, f.getInternalId()))
               .collect(Collectors.toSet()));
     }
 
-    GroupList orgGroups = _client.listGroups(generateGroupOrgPrefix(orgId), null, null);
-    if (orgGroups.stream().count() == 0) {
-      throw new IllegalGraphqlArgumentException(
-          String.format("Cannot add Okta user to nonexistent organization=%s", orgId));
+    // to remove...
+    Set<String> groupNamesToRemove = new HashSet<>(currentOrgGroupMapForUser.keySet());
+    groupNamesToRemove.removeIf(expectedOrgGroupNamesForUser::contains);
+
+    // to add...
+    Set<String> groupNamesToAdd = new HashSet<>(expectedOrgGroupNamesForUser);
+    groupNamesToAdd.removeIf(currentOrgGroupMapForUser::containsKey);
+
+    if (!groupNamesToRemove.isEmpty() || !groupNamesToAdd.isEmpty()) {
+      Map<String, Group> fullOrgGroupMap =
+          _client.listGroups(groupOrgPrefix, null, null).stream()
+              .filter(g -> GroupType.OKTA_GROUP.equals(g.getType()))
+              .collect(Collectors.toMap(g -> g.getProfile().getName(), g -> g));
+      if (fullOrgGroupMap.size() == 0) {
+        throw new IllegalGraphqlArgumentException(
+            String.format("Cannot add Okta user to nonexistent organization=%s", orgId));
+      }
+
+      for (String groupName : groupNamesToRemove) {
+        Group group = fullOrgGroupMap.get(groupName);
+        LOG.info("Removing {} from Okta group: {}", username, group.getProfile().getName());
+        user.addToGroup(group.getId());
+      }
+
+      groupNamesToAdd.stream()
+          .filter(n -> !fullOrgGroupMap.containsKey(n))
+          .forEach(
+              n -> {
+                throw new IllegalGraphqlArgumentException(
+                    String.format("Cannot add Okta user to nonexistent group=%s", n));
+              });
+      for (String groupName : groupNamesToAdd) {
+        Group group = fullOrgGroupMap.get(groupName);
```

Comment:
nit: would be a touch easier to read if you just added `if !fullOrgGroupMap.containsKey(groupName) throw new...` after line 289, and removed the stream right above